src/components/select/select.js

```diff
@@ -0,0 +1,340 @@
+        if (!options.find(selected => selected.id === selectedOption.id)) {
+          options.unshift({
+            id: selectedOption.id,
+            label: this.getOptionLabel(selectedOption),
+          });
+        }
+      });
+    }
+  }
+
+  componentDidMount() {
+    /* eslint-disable-next-line cup/no-undef  */
+    document.addEventListener('click', this.handleClickEvent, {
+      capture: true,
+    });
+  }
+
+  componentWillUnmount() {
+    /* eslint-disable-next-line cup/no-undef  */
+    document.removeEventListener('click', this.handleClickEvent, {
+      capture: true,
+    });
+  }
+
+  handleClickEvent = (event: MouseEvent) => {
+    // eslint-disable-next-line react/no-find-dom-node
+    const el = findDOMNode(this);
+    // $FlowFixMe
+    if (el && !el.contains(event.target)) {
+      this.setState({isDropDownOpen: false});
+    }
+  };
+
+  onFocus = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onBlur = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onMouseEnter = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onMouseLeave = (e: SyntheticEvent<HTMLInputElement>) => {};
+
+  onChange = (
+    e: SyntheticEvent<HTMLInputElement>,
+    type: ChangeActionT,
+    id?: string = '',
+    label?: LabelT,
+  ) => {
+    const multiple = this.isMultiple();
+    const selected = this.state.selectedOptions.find(tag => tag.id === id);
+    let selectedOptions;
+    switch (type) {
+      case STATE_CHANGE_TYPE.select:
+        if (!selected) {
+          selectedOptions = multiple ? this.state.selectedOptions.slice() : [];
+          selectedOptions.push({id, label});
+          this.setState({selectedOptions: selectedOptions});
+          if (this.props.type === TYPE.select && !multiple) {
+            this.setState({isDropDownOpen: false});
+          }
+          this.props.onChange(e, {type: type, id, label, selectedOptions});
+        } else if (multiple) {
+          selectedOptions = this.state.selectedOptions.filter(
+            selectedOption => selectedOption.id !== selected.id,
+          );
+          this.setState({
+            selectedOptions: selectedOptions,
+          });
+          this.props.onChange(e, {
+            type: STATE_CHANGE_TYPE.unselect,
+            id,
+            label,
+            selectedOptions,
+          });
+        }
+        break;
+      case STATE_CHANGE_TYPE.clearAll: {
+        const selectedOptions = [];
+        this.setState({selectedOptions: selectedOptions});
+        this.props.onChange(e, {type: type, selectedOptions});
+        break;
+      }
+      case STATE_CHANGE_TYPE.keyUp: {
+        // $FlowFixMe
+        const newTextValue = e.target.value;
+        this.setState({textValue: newTextValue});
+        if (this.props.simpleFilter) {
+          let filteredOptions = this.props.options.filter(
+            option =>
+              typeof option.label === 'string' &&
+              option.label.indexOf(newTextValue) >= 0,
```

Comment:
Can we do a case-insensitive match here? A lot of Selects have capitalized option names, whereas most users just enter a lowercase search query.

```
option.label.toLowerCase().indexOf(newTextValue.toLowerCase())
```

Reply:
sure